File: python/oa/quickfox.py
from collections import *

def get_words(text):
    return text.split(' ')
    
def solve(text, query, k):
    wt = get_words(text)
    wq = get_words(query)
    key = wq[0]
    
    ans = []
    
    n = len(wt)
    mp = defaultdict(int)
    for i in range(n):
        mp[wt[i]] += 1
        if i <= k: continue
        if wt[i-k] == key:
            ok = True
            for w in wq:
                if w not in mp:
                    ok = False
                    break
            if ok: ans.append(i-k)
        mp[wt[i-k]] -= 1
        if mp[wt[i-k]] <= 0: del mp[wt[i-k]]
            
    for i in range(n-1-k, n):
        if wt[i] == key:
            ok = True
            for w in wq:
                if w not in mp:
                    ok = False
                    break
            if ok: ans.append(i)
        mp[wt[i]] -= 1
        if mp[wt[i]] <= 0: del mp[wt[i]]
            
    return ans

text = "The quick brown fox is quick very very very quick fox"
query = "quick fox"
k = 2
ans = solve(text, query, k)
print(ans)

File: python/oa/quickfox.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> get_words(string s) {
    vector<string> r;
    string run = "";
    for (auto& x : s) {
        if (x == ' ') {
            r.push_back(run);
            run = "";
        }
        else run += x;
    }
    if (run != "") r.push_back(run);
    return r;
}

void solve(string text, string query, int k) {
    vector<string> wt = get_words(text);
    vector<string> wq = get_words(query);
    string key = wq[0];
    
    unordered_map<string, int> mp;
    int n = (int) wt.size();

    vector<int> ans;
    for (int i = 0; i < n; i++) {
        mp[wt[i]]++;
        if (i <= k) continue;
        if (wt[i-k] == key) {
            bool ok = true;
            for (auto& w : wq) {
                if (mp.count(w) == 0) {
                    ok = false;
                    break;
                }
            }
            if (ok) ans.push_back(i-k);
        }
        mp[wt[i-k]]--;
        if (mp[wt[i-k]] <= 0) mp.erase(wt[i-k]);
    }

    // process [n-1-k ... n]
    for (int i = max(0, n-1-k); i < n; i++) {
        if (wt[i] == key) {
            bool ok = true;
            for (auto& w : wq) {
                if (mp.count(w) == 0) {
                    ok = false;
                    break;
                }
            }
            if (ok) ans.push_back(i);
        }
        mp[wt[i]]--;
        if (mp[wt[i]] <= 0) mp.erase(wt[i]);
    }

    for (auto& x : ans) cout << x << ' ';
}

int main() {
    // string text = "The quick brown fox is quick very very very quick fox";
    // string query = "quick fox";
    // int k = 2;
    string text; getline(cin, text);
    string query; getline(cin, query);
    int k; cin >> k;
    solve(text, query, k);
}